python/cudf/cudf/core/column/column.py

```diff
@@ -901,7 +901,10 @@ def as_categorical_column(self, dtype, **kwargs) -> ColumnBase:
         sr = cudf.Series(self)
 
         # Re-label self w.r.t. the provided categories
-        if isinstance(dtype, (cudf.CategoricalDtype, pd.CategoricalDtype)):
+        if (
+            isinstance(dtype, (cudf.CategoricalDtype, pd.CategoricalDtype))
+            and dtype._categories is not None
```

Comment:
For `pd.CategoricalDtype`, `dtype.categories` would be a more "public" way to access the categories, but understand if this makes it generic with `cudf.CategoricalDtype`

Reply:
@galipremsagar I figured the reason we  were using `_categories`  was to avoid materializing it unnecessarily, but it looks like we call `dtype.categories` below anyway. Is there any advantage to using `_categories` here  instead?

Reply:
So the reason we had to use `._categories` is, when there is an **empty** categorical type created, cudf has the following issue where all the values are converted to `<NA>`:
```python
In [1]: import cudf

In [2]: cat_type = cudf.CategoricalDtype()

In [3]: s = cudf.Series([1, 2, 3])

In [4]: s
Out[4]: 
0    1
1    2
2    3
dtype: int64

In [5]: s.astype(cat_type)
Out[5]: 
0    <NA>
1    <NA>
2    <NA>
dtype: category
Categories (0, object): []

In [7]: s.to_pandas().astype(cat_type.to_pandas())
Out[7]: 
0    1
1    2
2    3
dtype: category
Categories (3, int64): [1, 2, 3]
```

Whereas pandas handles this correctly i.e., a `pd.CategoricalDtype(None, None)` is the same as a string `"category"`. Inspecting `._categories` is essential to distinguish and fix this issue because `cudf.CategoricalDtype.categories` returns an empty GPU object instead of `None`.

Reply:
Got it -- so can we only inspect `._categories` in the cuDF case and `.categories` in the Pandas case?

Reply:
Yup, and thus had to access `._categories` to keep it simple as that works for pandas case too.

Reply:
Let's split out the Pandas logic to use `.categories`  as there's no guarantee Pandas will always have a `._categories` attribute?

Reply:
Sure, made the separation. Mypy also exposed a bug 😉 and fixed it too.